Dialog interface guidelines

langermank opened the pull request:
Interface guidelines for the updated Dialog component coming soon to PVC! ✨ 

See preview deploy: https://primer-f643fd6630-26441320.drafts.github.io/components/dialog

<img width="960" alt="Dialog anatomy visual with internal labels pointing to each region" src="https://user-images.githubusercontent.com/18661030/185509794-0a3b578e-40b2-43db-ab61-8731c9888702.png">

This is a scoped down version of the Overlay interface guidelines #167. We decided that these guidelines are best suited as component specific compared to a general pattern overview. After Dialog, we'll tackle other Overlay related `ActionMenu`, `SelectPanel`, `SelectMenu` and update `Autocomplete` interface guidelines which will cover the specifics of those components.

**Feedback**

- [ ] Check spelling 🙌 
- [ ] Are the corresponding images clear? Is each caption sufficient?
- [ ] Is there anything else you want to see included here?

PVC implementation: https://github.com/primer/view_components/pull/1214
Closes https://github.com/github/primer/issues/656

jonrohan commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+
+## Anatomy
+
+<img width="960" alt="" src="https://user-images.githubusercontent.com/18661030/185509649-50791e2f-0a4c-44d9-8b2b-e34b3f394315.png">
```
```suggestion
<img width="960" alt="" src="https://user-images.githubusercontent.com/18661030/185509649-50791e2f-0a4c-44d9-8b2b-e34b3f394315.png" />
```

langermank commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+
+The `height` and `width` may be set to `auto` which will set a minimum width of `192px` and respond to the contents of the Dialog. Additionally, an explicit `height` and `width` may be set with a specific set of options.
+
+<img width="1292" alt="" src="https://user-images.githubusercontent.com/18661030/185178528-e1c176ef-f407-40ab-8729-965e21d64f6d.png"/>
```
I'm curious how folks feel about this image. We offer a large amount of size combinations, and finding a way to show that in a static image is tricky! More discussion [here](https://github.com/github/primer/pull/1202/files#r944065171). Open to suggestions!

langermank replied:
Here are some alternative options:

<img width="1292" alt="Size options for Dialog" src="https://user-images.githubusercontent.com/18661030/185514743-c27965cb-08f6-4236-8cec-663788e1f2dc.png">

This ones from @vdepizzol!

<img width="1216" alt="Size options for Dialog" src="https://user-images.githubusercontent.com/18661030/185514800-c2d231aa-11fb-4f51-a23c-4c54649ce0b3.png">

josepmartins replied:
This is great ✨ some thoughts/comments: 

- 👍  to add a width/height label to the X/Y axis
- 👍  to use a single purple color rather than the sequential palette. It's more clear when trying to understand what is a ✅ and a 🔴 combination.
- What about a legend along with the image with all the possible size combinations?
- Should we explain why we provide these combinations and not others? e.g, line length good practices

josepmartins replied:
This is great ✨ some thoughts/comments:

- 👍 to add a width/height label to the X/Y axis
- 👍 to use a single purple color rather than the sequential palette. It's clearer when looking for what is a ✅ and a 🔴 combination.
- What about a legend along with the image with all the possible size combinations?
- Should we explain why we provide these combinations and not others? e.g, line length good practices

vdepizzol replied:
I think this has been solved with the newer sizing options? Marking as resolved.

lukasoppermann commented:
Hey @langermank,

This is really awesome. After reading through it feels like I know everything there is to know about the Dialog. ❤️
Also the illustrations are on point and everything important aspect is visually covered as well.

I have a few little points. Please excuse me if they are not valid / clear to everyone else, as I am still kinda new to this. 😉 
Also if this is too detailed or nitpick-y please let me know for next time. 😅

### General
- [ ] Maybe we could simplify the intro? 
(_Dialogs create a new modality to the user, and can facilitate the completion of individual tasks._) 
This may be a bit hard to understand for non-native speakers. Especially "new modality" and "facilitate" could maybe be replaced by a simpler alternative?
### Header
- [ ] The close button in the "Anatomy of a Dialog" picture seems to be misaligned, as it is centered with the title in all other screens.
Same here: https://primer-f643fd6630-26441320.drafts.github.io/components/dialog#keep-content-simple I get the feeling I just don't get the reason. 😅
- [ ] Question: It states that for the command palette there has to be a close button as well. Currently the input in the command palette has a clear button (x within circle). Has this been considered? We would have the clear x and the close x very close to each other which could lead to confusion. Just wondering if that was considered and deemed fine.
### Footer region
- [ ] since the divider is just a thin line, it may be helpful to point to it with a purple annotation line in the first picture were you talk about the optional divider
- [ ] It could be helpful to link to the [button alignment section](https://primer-f643fd6630-26441320.drafts.github.io/ui-patterns/button-usage#alignments) or even add a small text as to why "cancel" and the primary button are both right aligned, but the "New label" button is left aligned
- [ ] could "auxiliary" be replaced with a simpler alternative in "Footer with auxiliary create button"?

langermank commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+
+### Backdrop
+
+Dialog displays with a **backdrop**, which dims the rest of the page. The backdrop is visible when the Dialog is centered on the page, or attached to a sides as a sheet.
```
```suggestion
Dialog displays with a **backdrop**, which dims the rest of the page. The backdrop is visible when the Dialog is centered on the page, or attached to a side as a sheet.
```

alliethu commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+
+### Trigger elements and keyboard focus
+
+Dialogs are initially hidden, and can be opened by an element with `role="button"`, called a **trigger**. The role `button` allows the **trigger** to be a relativelly complex element, such as an action list item.
```
Misspelling of `relatively`

alliethu commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+
+Dialogs are initially hidden, and can be opened by an element with `role="button"`, called a **trigger**. The role `button` allows the **trigger** to be a relativelly complex element, such as an action list item.
+
+Dialogs should be fully functional using the keyboard and assistive technology. Pressing `Esc` should dismiss a Dialog, also returning focus to the **trigger** that opened it.
```
This `should` language makes it seem optional vs. using `must`. Thoughts on adjusting?

langermank commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+
+### Trigger elements and keyboard focus
+
+Dialogs are initially hidden, and can be opened by an element with `role="button"`, called a **trigger**. The role `button` allows the **trigger** to be a relativelly complex element, such as an action list item.
```
```suggestion
Dialogs are initially hidden, and can be opened by an element with `role="button"`, called a **trigger**. The role `button` allows the **trigger** to be a relatively complex element, such as an action list item.
```

langermank commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+
+Dialogs are initially hidden, and can be opened by an element with `role="button"`, called a **trigger**. The role `button` allows the **trigger** to be a relatively complex element, such as an action list item.
+
+Dialogs should be fully functional using the keyboard and assistive technology. Pressing `Esc` should dismiss a Dialog, also returning focus to the **trigger** that opened it.
```
```suggestion
Dialogs must be fully functional using the keyboard and assistive technology. Pressing `Esc` should dismiss a Dialog, also returning focus to the **trigger** that opened it.
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+Side sheets take the entire height of the viewport, and should be used sparingly. Don't use side sheets to present forms or flows that should be a page instead.
+
+<img width="960" alt="placement-desktop-1" src="https://user-images.githubusercontent.com/18661030/185214096-8ad908b4-2cbd-467d-8e74-bad6948c5a2d.png"/>
+<Caption>Left aligned side sheet for global navigation.</Caption>
```
```suggestion
<Caption>Left aligned side sheet is reserved for global navigation only.</Caption>
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+
+Side sheets slide from the right or left edges of the viewport. **Left side sheets are reserved for global navigation drawers. Right side sheets are used for global actions, but can also be considered for quick previews in full-width pages, or configuration panels.**
+
+Side sheets take the entire height of the viewport, and should be used sparingly. Don't use side sheets to present forms or flows that should be a page instead.
```
Trying to make it a little clearer when it "should be a page instead":

```suggestion
Side sheets take the entire height of the viewport, and should be used sparingly. Don't use side sheets to present create/edit forms, or flows that main contain a lot of information. For that, use a page instead.
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+<Caption>Left aligned side sheet for global navigation.</Caption>
+
+<img width="960" alt="placement-desktop-2" src="https://user-images.githubusercontent.com/18661030/185214094-419d2271-dbb9-4e78-90e6-4433ed9c42fb.png"/>
+<Caption>Right aligned side sheet for global actions.</Caption>
```
```suggestion
<Caption>Right aligned side sheet can be used for global actions, but may also be considered for quick previews or configuration panels.</Caption>
```

vdepizzol reviewed (commented):
Trying to make the explanation for side sheet usage a little clearer :))

joshblack reviewed (commented):
Looks great! 🥳 The images are fantastic, I love them so much.

**Feedback checklist**

- [x] Check spelling 🙌 
- [x] Are the corresponding images clear? Is each caption sufficient?

Love the images, I think the dimensions one was the only one that was confusing initially (I think there is a thread above that's already talking about other ways to approach this)

- [x] Is there anything else you want to see included here?

I think the only thing that came to mind was around keyboard interaction and focus behavior for the different areas. Specifically:

  - Where should focus be placed when the dialog is open (is this configurable?)
  - In what situations is the dialog modal or nonmodal? In nonmodal situations, is focus still trapped?

Similarly, is it worth calling out when content outside of the dialog is inert? (This might already be covered by the backdrop, I wasn't sure)

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,265 @@
+    <Caption>Body content will scroll vertically if the content exceeds the overall Dialog height.</Caption>
+  </div>
+  <div>
+    <img width="650" alt="" src="https://user-images.githubusercontent.com/18661030/185478320-2a4a3053-dca3-4048-b7c4-d5a39a83acd6.png" />
+    <Caption>The body may contain additional content such as banners and form controls.</Caption>
```
I think we can simplify this example a bit:

Before:

> ![image](https://user-images.githubusercontent.com/293280/185690249-94e18546-9275-43e3-a5a7-28783a1c787f.png)

After:

> ![image](https://user-images.githubusercontent.com/293280/185690317-526f4baf-4cfc-40ca-8587-9a0c64d5bd6a.png)

```suggestion
    <img width="650" alt="" src="https://user-images.githubusercontent.com/293280/185690317-526f4baf-4cfc-40ca-8587-9a0c64d5bd6a.png" />
    <Caption>The body may contain additional content such as banners and form controls.</Caption>
```

langermank replied:
Thanks @vdepizzol! I updated the image to be closer to your recommendation. The height/width combo in the one you shared is in our disallowed list and I want this to be a little more accurate if possible!

langermank commented:
@lukasoppermann thanks for the feedback! Addressing your questions here:

> The close button in the "Anatomy of a Dialog" picture seems to be misaligned, as it is centered with the title in all other screens.

The button alignment is intentional– you can see better in this graphic that the icon inside the button is visually aligned with the Dialog title. I wanted the graphics to be somewhat accurate, but I wonder if its too distracting.

![Dialog title with close button](https://user-images.githubusercontent.com/18661030/185988836-16de60d4-c2e8-462a-941d-be815cbfd66a.png)

> Command palette 

I'm not sure that anything has been finalized for this, but I wanted to include an exploration here. I'm basing it off of an audit I found from @ichelsea. I just added a "Clear" button to the input to at least make this exploratory graphic a bit more realistic!

![Dialog as a command palette](https://user-images.githubusercontent.com/18661030/185989873-9ec5a64f-0c86-4ae3-9e8e-c595906f0d6d.png)

> Footer line

Not sure where you mean, I think I always highlight the divider lines in all the images.

> Footer buttons / language

Addressed!

langermank commented:
@joshblack thanks for the feedback! Addressing your questions here:

> Where should focus be placed when the dialog is open (is this configurable?)

I believe it will focus the first interactive item, which will always be the close button since we don't allow it to be hidden. This is visible in the videos, but I'm going to repeat this in an image!

> In what situations is the dialog modal or nonmodal? In nonmodal situations, is focus still trapped?

I'm going to hold off on addressing this one. For `Dialog` I think we're always assuming `role="dialog"`. However, we have a few specialized components in the pipeline that also have that role but behave slightly differently (like a menu type dialog that's anchored to a trigger). 

My overall approach for the accessibility concerns was to kind of bake it into the content, but I decided to go ahead and add another section for it to just call out focus trapping specifically. Its somewhat repeated content but I don't see any harm in reiterating how Dialog should behave!

josepmartins commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,383 @@
+  alt=""
+  src="https://user-images.githubusercontent.com/18661030/185178528-e1c176ef-f407-40ab-8729-965e21d64f6d.png"
+/>
+<Caption>Recommended height and width combinations for Dialog</Caption>
```
+1 to the table or even a legend with all the possible size combinations:

Other thoughts on this regarding this [comment](https://github.com/primer/design/pull/281#discussion_r949690844):
- 👍 adding a width/height label to the X/Y axis
- 👍 to use a single purple color rather than the sequential palette. It's clearer when looking for what will be a good or a bad combination.
- Should we explain why we provide these combinations and not others? e.g, line length good practices

josepmartins commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,383 @@
+
+Dialog's title and close button as addressed previously are required to ensure an accessible experience for everyone. Further, keyboard and focus behavior must be considered.
+
+When a Dialog is opened, the first interactive element (the close button) is focused. Hitting `tab` will cycle through all interactive elements within the Dialog. To escape the focus trap, hitting the `esc` key or clicking on the backdrop will close the Dialog. While the Dialog is open, page scrolling is disabled and becomes `inert`.
```
Does the focus behavior change when using a search field in the `Subheader` region or a Command palette in the `Header` region?

<img width="505" alt="Screenshot 2022-08-25 at 11 35 27" src="https://user-images.githubusercontent.com/912236/186631087-664c4836-3429-4cf5-841f-1d63952746f5.png">

langermank replied:
Great question! Yes, since the input comes before the close button in the dom it should receive focus first. I added this to the accessibility section!

langermank commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,383 @@
+
+### Subheader region
+
+The subheader region is an optional space for interactive controls. Use it to display a search field, a filter menu, or a local navigation component.
```
I'm not sure, honestly. I think I might leave this as is for now, and if we learn of something that's not allowed update it.

langermank commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,383 @@
+    <img
+      width="650"
+      alt=""
+      src="https://user-images.githubusercontent.com/18661030/186003712-36441f05-1407-4b03-a50a-3773275e4d61.png"
```
Did it stand out to you as odd? It stood out to me. I think I'll change it.. if we want alerts to look a specific way in Dialog, we should consider offering a special slot for it 🤔

mperrotti replied:
Yea, it stood out as odd to me too.

Agreed with offering a special slot instead of manually applying different styling.

langermank commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,383 @@
+  <div>
+    <CustomVideoPlayer
+      width="100%"
+      src="https://user-images.githubusercontent.com/18661030/185507523-74e0acff-6980-439e-84c5-422b16470c9e.mp4"
```
Yes! I used this plugin to make it: https://www.figma.com/community/plugin/824289601590456356

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+  alt=""
+  src="https://user-images.githubusercontent.com/18661030/187482408-97bcb753-62a9-4b7c-b02b-b94ba369203e.png"
+/>
+<Caption>Small Dialogs have a min width of 320px, and a max height of 256px.</Caption>
```
> have a min width of 320px

I'm seeing these descriptions across the different dialog sizes, but doesn't the "min width" makes it sound like the dialog won't get smaller than that? All dialogs will shrink on smaller viewports.

langermank replied:
In the main description I mention: 

> Dialogs will never extend beyond the viewport, and the body content must fit a narrow viewport with a minimum width of 320px.

Is it not clear that the dialog will always retract to fit inside the viewport? My goal is that a designer reading this can get a sense of sizing, while also understanding that the size is variable depending on the available screen space.

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+  4. Footer buttons always appear inside the safe area. <br />
+</Caption>
+
+## Size
```
Did we remove support for `auto` height?

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+
+Dialog offers a range of sizes from `small` to `xlarge` with a default of `medium`. Each size provides a `max-height` and will automatically shrink to its body content. Dialogs will never extend beyond the viewport, and the body content must fit a narrow viewport with a minimum width of `320px`.
+
+### Small
```
What's the use case of having a max-height of 256px in the `small` size?

langermank replied:
Are you suggesting it should have a different max-height, or no max-height? Its following the same pattern as all sizes to maintain a ratio. I'm offering a small portrait mode as well, but that's only for menu scenarios. @vdepizzol

vdepizzol replied:
@langermank I'm trying to picture which scenarios we'd like to have a landscape `small`, with the 320px x 256px dimensions... 

For confirmation dialogs, the 320px width is used but I wouldn't expect it to cap the height at 256px, for example — in that case it'd keep an `auto` height if there's a lot of text in it.

In case of a `small portrait` menu, would it cap the height at a certain size and force scrolling even if there's space?

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,383 @@
+      src="https://user-images.githubusercontent.com/293280/175850637-3b193552-b35f-4b07-b2d6-4ca546c8131a.png"
+      alt=""
+    />
+    <Caption>Footer with a secondary button aligned to the left.</Caption>
```
Left-aligned buttons can be used for auxiliary actions. Maybe we can use that wording in there?

```suggestion
    <Caption>Footer with an auxiliary action aligned to the left.</Caption>
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+
+### Extra large
+
+Before using an extra large dialog, consider whether or not the content would be more appropriate for a new page.
```
Capitalizing the component name:

```suggestion
Before using an extra large Dialog, consider whether or not the content would be more appropriate for a new page.
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+  alt=""
+  src="https://user-images.githubusercontent.com/18661030/185214079-d30342c1-6061-4a31-ab65-c0cef0f00278.png"
+/>
+<Caption>Full-screen dialog in portrait mode.</Caption>
```
Capitalizing the component name:

```suggestion
<Caption>Full-screen Dialog in portrait mode.</Caption>
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+  alt=""
+  src="https://user-images.githubusercontent.com/18661030/187482817-9d82fe3e-c24d-4eca-9b4b-c74fa32450d4.png"
+/>
+<Caption>Extra large Dialogs have a min width of 960px, and a max height of 600px.</Caption>
```
```suggestion
<Caption>Extra large Dialogs have a width of 960px, and a max height of 600px.</Caption>
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+  alt=""
+  src="https://user-images.githubusercontent.com/18661030/187482693-04eb3780-89e1-4705-a55b-5fa8ba0f9865.png"
+/>
+<Caption>Large Dialogs have a min width of 640px, and a max height of 432px.</Caption>
```
```suggestion
<Caption>Large Dialogs have a width of 640px, and a max height of 432px.</Caption>
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+  alt=""
+  src="https://user-images.githubusercontent.com/18661030/187482576-557672f9-6695-435d-b93e-2b7b8fda2a98.png"
+/>
+<Caption>Medium Dialogs have a min width of 480px, and a max height of 320px. Medium portrait oriented Dialogs have a min width of 480px, and a max height of 600px.</Caption>
```
```suggestion
<Caption>Medium Dialogs have a width of 480px, and a max height of 320px. Medium portrait oriented Dialogs have a width of 480px, and a max height of 600px.</Caption>
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+  alt=""
+  src="https://user-images.githubusercontent.com/18661030/187482408-97bcb753-62a9-4b7c-b02b-b94ba369203e.png"
+/>
+<Caption>Small Dialogs have a min width of 320px, and a max height of 256px.</Caption>
```
```suggestion
<Caption>Small Dialogs have a width of 320px, and a max height of 256px.</Caption>
```

vdepizzol commented on content/components/dialog.mdx:
```diff
@@ -0,0 +1,420 @@
+
+## Size
+
+Dialog offers a range of sizes from `small` to `xlarge` with a default of `medium`. Each size provides a `max-height` and will automatically shrink to its body content. Dialogs will never extend beyond the viewport, and the body content must fit a narrow viewport with a minimum width of `320px`.
```
Trying to make this easier to parse:

```suggestion
Dialog offers a range of sizes from `small` to `xlarge`, with the default size set to `medium`. 

Each size defines a width and a maximum height. By default the dialog height will adjust to the body content. If the maximum height is reached, the body contents will scroll.

A Dialog sizing is constrained by the viewport dimensions. Dialogs will not grow beyond the boundaries of the viewport.

Small viewports force Dialogs to remain small. All Dialogs need to work on viewports of at least `320px` of width, and `256px` of height.
```

langermank replied:
Beautiful! Thank you 🙌

mperrotti reviewed (approved):
Looks great! I love the new sizing content.

vdepizzol reviewed (approved):
Great work, @langermank! Excited to see this live! Thank you everyone for reviewing 🎉 🎉
